src/Controls/src/Build.Tasks/CompiledMarkupExtensions/OnIdiomExtension.cs

```diff
@@ -0,0 +1,113 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Xml;
+using Microsoft.Maui.Controls.Xaml;
+using Mono.Cecil;
+using Mono.Cecil.Cil;
+using static System.String;
+using static Mono.Cecil.Cil.Instruction;
+using static Mono.Cecil.Cil.OpCodes;
+
+namespace Microsoft.Maui.Controls.Build.Tasks;
+
+class OnIdiomExtension : ICompiledMarkupExtension
+{
+	public IEnumerable<Instruction> ProvideValue(IElementNode node, ModuleDefinition module, ILContext context, out TypeReference memberRef)
+	{
+		//if all idiom values have the same type, we can return that type for memberRef, otherwise we default to object
+		memberRef = module.TypeSystem.Object;
+
+		var visitor = new SetPropertiesVisitor(context, stopOnResourceDictionary: true);
+		foreach (var cnode in node.Properties.Values.ToList())
+		{
+			if (!(cnode is IElementNode enode))
+				continue;
+			foreach (var jnode in enode.Properties.Values.ToList())
+				jnode.Accept(visitor, jnode);
+			foreach (var jnode in enode.CollectionItems)
+				jnode.Accept(visitor, jnode);
+		}
+
+		var parentNode = node.Parent;
+		if (!SetPropertiesVisitor.TryGetPropertyName(node, parentNode, out var propertyName))
+			return null;
+
+		var idioms = new[] { "Phone", "Tablet", "Desktop", "TV", "Watch", "Default" };
+		var idiomProperties = node.Properties.Where(p => idioms.Contains(p.Key.LocalName)).ToList();
+
+		//assign memberRef to the type of the first idiom value, then compare with the other ones
+		memberRef = (idiomProperties.FirstOrDefault().Value as IElementNode).XmlType.GetTypeReference(context.Cache, module, node as IXmlLineInfo);
+		foreach (var idiom in idiomProperties.Skip(1))
+		{
+			var idiomTypeRef = (idiom.Value as IElementNode).XmlType.GetTypeReference(context.Cache, module, node as IXmlLineInfo);
+			if (TypeRefComparer.Instance.Compare(memberRef, idiomTypeRef) != 0)
+			{
+				//should this issue a warning?
+				memberRef = module.TypeSystem.Object;
+			}
```

Comment:
I'm worried this comparison is too strict, especially when comparing ints and doubles (e.g., `Phone=0, Tablet=2.5`). Would it be possible to try to find the "lowest common denominator" between all the types instead of immediately fall back to `object`?

Reply:
other option is to use the target type, and convert

Reply:
I added a test for that, and we got the right behaviour, even with fallback to object

Reply:
that doesn't work :/